### ISSUE TITLE
Identity Conditions

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#32]
> _Original title: Conditions and Entity Arrays_
### API

During the work on the post terms Condition, the Entity API was born (originally christened the Args API). One part of it was the concept of Entity Arrays (https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125251112). This was needed for representing a post's terms.

The possibility of Conditions for Entity Arrays was explored early on in the context of the API (https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125249274  and https://github.com/WordPoints/hooks-api/issues/13#issuecomment-125251112). These would be necessary in order to, for example, require that a post have a given term. Normally Conditions are related to a specific attribute of a specific Entity in the hierarchy, for example, a post's content. This is simple, because a post only has one content attribute. But we needed to have a way to add a Condition on a set of items related to an Entity, like a post's terms.

Because we don't know what the Entity Array will contain beforehand, and Conditions have to be applied to named sub-args, it was decided that Conditions could only be applied to the elements within an Entity Array by applying a Condition to the Entity Array collectively.

So for example, if we want a Condition that requires a post to have a given term, what we are really saying is that we want to require the array of post terms to include that given term. And that is how it is treated in the API.
### UI

Later, during development of the Conditions UI, we had to decide how Entity Arrays should be presented. 

One part of the UI consists of a set of dropdowns which are used to select the Entity child to add a Condition on. The first dropdown contains a list of the args for that Event. Then once one of the args is selected, a new dropdown is displayed listing its children. This process is continued until the arg in the last dropdown doesn't have any children to display. Then a dropdown showing the list of available Conditions is shown, based on what type of arg has been selected.

When a Condition for an Entity Array is added, the user may (depending on the condition) be given the opportunity to repeat the process of selecting some Conditions to add within the Entity Array. In other words, they can select conditions that at lest x of the items in the array must match.
### UI

There are problems here, mostly in terms of the UI. 

One of the biggest issues is that the current UI is inelegant. To require that a user have a given role for example, you have to select User » Roles, then select the Contains condition. After adding the condition, we then have to add add sub-condition. This condition will have to be on the ID for the entity, which in the case of the user roles is a slug, called the "name". So we have to select the Name arg, and then the Equals condition. The sub-condition then offers a dropdown with the list of available user roles. Of course, we don't display the IDs in that dropdown, instead we display human readable IDs.

For the user roles, this doesn't work out too bad, because the Name arg makes sense. However, if we ever wanted to let the user add a condition on the actual title of the entity (the human-readable ID shown in the "name" slug dropdown), the user would probably be confused.

For something like a post's terms, the ID of the term Entity is called the `term_id`, or possibly the "ID". But this seems strange, to be showing the user the ID arg in the dropdown, and then when they select the ID, showing them the list of term names (I assume that we'd do something like that, although how to handle that hasn't been finalized yet). The point is that the user will be filtering an arg named ID, but we probably won't be asking them to actually enter a numeric ID. IDs are, after all, meant for the code, not for users.

The question here is how we can better handle this in the UI.
### API

In the API we have a related issue: we have to create entity classes and register entity children for the ID fields of each Entity that will be used in an Entity Array like this. However, that seems odd, because we already have the data we need (`$id_field`, etc.) within the Entity class itself. The only reason that we need to register the child is for it to be displayed in the UI, but as pointed out above, it is odd that we'd ever need to do that. In short, having to add the children seems like unnecessary duplication.

**Comment by JDGrimes** [comment]
## Broader Scope

I guess this really goes beyond entity arrays and touches all entities which we might want to have conditions on the ID attribute. One example would be post types (but see #29). How do we allow a user to set a condition on the identity of the entity, without overexposing the ID field or non-human readable IDs in the UI?

## Confusion for Users

The concern that I brought up above is that if we call the field "ID", and display a list of human-readable IDs which are derived from other fields, this might confuse the user. Actually, it might confuse the user in two different ways. 
1. First, it is confusing because it seems to be duplication (tag Name and tag ID both seem to be the same thing). 
2. Second, it is confusing because the user might start to think that the human-readable ID itself is what actually being used to identify the field internally, when it isn't.

The significance of that second one is that when the human ID changes, the internal ID remains the same.

I had actually considered that one solution would be to use the human ID as the value we actually save to run the condition against. That way we could eliminate the ID field from the UI completely. But of course then the condition would cease to work as (would probably be) expected if/when the human ID field was changed (which is like a post title, tag name, etc.). Actually, that could probably even be broken by l10n related stuff, much more easily than the actual ID.

## Considerations with using "ID" in the UI

I'm actually not as uncomfortable as I thought I might be with using the term ID or identifier in the UI. However, we may want to be careful about standardizing the name of the identifier in the UI, for two reasons:
1. It could possibly conflict with the name of other attributes for some attributes.
2. It might confuse users if the entity ID is clearly identified by a different name somewhere else in some UI.

**Comment by JDGrimes** [comment]
## UI: Direct Identity Conditions

It might seem that our problems would be solved if we just let the identity conditions be directly on the entity in the UI by having the condition selector appear when the entity was selected. The problem is that the entity may (likely _will_) also have children. So we have to choose between showing two dropdowns, one with the children and one with the identity conditions, or displaying both identity conditions and children in a single dropdown. Both have their problems.

### Two Dropdowns

#### Problems
- Complicates the UI.
- Could be confusing.

This probably could actually be done without making the UI too complex or confusing if both of the selectors were labeled.

`
{Select...} or {Select condition...}
`

### Dual-purpose Dropdown

#### Problems
- Might be ambiguous.
- Or at least rather confusing.
- Could be hard to choose a good label (needed for a11y).

The first two in that list could be overcome in several ways, like prefixing the condition names with "Condition: ".

The label wouldn't necessarily be the shortest on earth, but could probably be pieced together without much trouble.

---

However, there are also other issues with direct identity conditions. Like entity arrays, which is why I originally connected this issue specifically to entity arrays.

When we have a condition on the entity array, we have the option to add sub-conditions. When the entity has the potential for identity conditions, we'd run into the same conundrum as above. But I guess it could be solved in the same ways.

**Comment by JDGrimes** [comment]
There is also a more general problem with direct identity conditions. When we're talking about the Equals condition, it is one thing. But what if the user has legitimate reasons for wanting to add other types of conditions on the ID attribute? At that point, adding the condition directly on the entity really doesn't make sense. So we'd have to register the ID attribute as an entity child, and that would confuse things even further.

Basically, direct identity conditions are an attempt to cater to a particular condition specifically, the Equals condition. But that is really the only condition for which it is of value.

**Comment by JDGrimes** [comment]
## API: Registering ID Atts

Registering the ID atts of each entity does seem like duplication, since this information is already defined in the Entity class itself. However, I don't think that it is unwarranted, because the entity ID object has the opportunity to do much more than just define the field name, it can also define information about what type of field it is (integer, slug), etc. 

Of course, it isn't needed for all entities either, only those which we want to be able to place conditions on the ID of, regardless of how those conditions are presented in the UI.

**Comment by JDGrimes** [comment]
Another problem with just placing the identity equals condition on the human ID would be that not all Entities provide such a human ID which would really make sense here (see #28).

## The Equals Condition _is_ Special

> Basically, direct identity conditions are an attempt to cater to a particular condition specifically, the Equals condition. But that is really the only condition for which it is of value.

I don't think it is actually so far-fetched to want to do this. The identity equals condition is special, and has been presented in this fashion in the past (the Post Type field on the Post Publish points hook was a dropdown with the post type names [slugs] as values and the labels as the option texts).

However, we also need to consider whether we are catering to a specific subset of entities as well: those that are enumerable, like the post type or user role. At run-time we have a finite list of registered post types that are available on the site. There usually isn't going to be more than a few handfuls of them, so it is very easy to display them in a dropdown in this way (and we don't even have to run a query to get the data!). But what about an entity like a Post or Term, that is queryable instead? There really isn't any limit to how many there can be, and displaying them in a dropdown isn't practical. Although the query API isn't finalized yet, I envision us offering hints in the UI instead.

In other words, the identity equals condition is special because it needs to change how it presents the UI, not based just on the data type of the ID, but, perhaps, on how the entity is stored. Though I guess, in fact, the enumerable entities could use hinting just like the queryable entities, they just wouldn't have to send off a query to the sites database to do so.

While writing that I happened to think that maybe there are other fields besides the ID on which we might want to use the equals condition, _and_ offer hinting. Of course there are other places that the equals condition would be useful, like if we want a post meta field to equal a certain value, but I guess that even there we could offer hinting. So maybe this is something specific to the equals condition, but not just the _identity_ equals condition.

(Then again, perhaps there are some cases, like the post publish date, where hinting for an equals condition wouldn't be useful. Though that would match down to the second anyway, so the whole condition wouldn't be of much use. Anyway, we could let that UI feature be determined by the handler for each data type.)

## Questioning Identity Conditions

I think maybe we need to ask a much more basic question: should we have identity equals conditions at all?

If we want to only award points for [viewing/commenting on/etc.] a specific post, isn't that much better managed on the post edit screen than as a reaction? More generally, doesn't it make more sense, if a reaction is tied specifically to a single entity, to manage it alongside that entity instead of with the list of general reactions on the points types screen? (Though it wouldn't have to use the hooks API, necessarily.) And if this basically holds true for all entities (although I wonder whether there would actually be present a UI for all of them, but I digress), couldn't we avoid ever having identity conditions at all?

Yes and no.

Unfortunately, at best, this only solves half of our problem. We would no longer have to have identity conditions for the args of the event. But what of the args they are related to? Which brings us back to the example of the post-has-term condition. We aren't setting the identity condition on the post but on a related entity (which in this example happens to be in an array).

To avoid that, we could set the condition on an attribute of the primary entity instead of the ID of the related entity. For example, instead of setting the attribute on the post type name (slug/ID) we could set it on the `post_type` field of the post. This is basically what we did before in the Post Publish points hook.

There are two problems with it, however.
1. It doesn't let us set conditions on other atts (or children, generally) of the related entity (unless we duplicate things by displaying both the field and the relationship in the arg children dropdown).
2. It only works for relationships where the ID field of the related entity is also the field of a field on the primary entity (though I think we could get around that).

In short, this is basically the same as playing with the dropdowns. The difference between this and adding the identity field to the list of children, is that the ID field is basically displayed at a higher level instead. I seem to recall considering that possibility before, when first building the UI, but because of the problems with it I didn't pursue it.

**Comment by JDGrimes** [comment]
One of the side effects of there being just one identity condition is that we would only have to display one in the list of child args if we go the mixed dropdown route.

Then again, I'm not so sure that the identity equals condition is the only identity condition. I'm not talking about users wanting to set other conditions on the ID field, which is different than setting an identity condition. I'm thinking really that if we have `Post Type » Equals` we might also want other things, like `Post Type » Not Equals`. Or even other things like `Post Type » Is {something}`, which could be supplied by a dynamic function under the hood.

I was just thinking that there might also need to be the Contains condition as an identity condition for one-to-many relationships, like post Terms. But I was actually confused. The Contains condition would exist just as it does now, it wouldn't be an identity condition. The array itself doesn't have an identity, though identity conditions could be added as sub-conditions.

**Comment by JDGrimes** [comment]
Perhaps we need to consider re-imagining the UI completely.

**Comment by JDGrimes** [comment]
> Second, it is confusing because the user might start to think that the human-readable ID itself is what actually being used to identify the field internally, when it isn't.

One possible solution for this would be to also show the ID of the entity in the dropdown (like `Human ID (ID)`). But then that makes the issue of exposing the ID in the API even worse, because we're now not just using the term "ID" but also the ID itself. I think that really I'd be more comfortable with the potential for confusion than with this.

However, I just happened to think that if/when we introduce the placeholders feature to the logs API, that kind of confusion could actually be significant. Because a confused admin might end up referencing the ID in the log text instead of the Human ID.

**Comment by JDGrimes** [comment]
## How Users Perceive Identity

Regardless of how we present it in the UI, we are going to have to express the concept of Identity to the user in some way. The question is how best to do this.

The fact is that the identity of an entity is determined by one of its attributes. Usually this is also an attribute that the user seldom (or never) sees.

When a user thinks of a particular tag, for example, they don't think of ID 897. They think "Snow", or "Horses". That is the Human ID, and what it is called differs from one type of entity to the next.

The user doesn't just think of the Human ID as the "title" of the tag, or even the ID of the tag; to them, it _is_ the tag. For the user, the concept of identity isn't so much an attribute of the entity as it is the representative of the entity.

So, for the user, it might seem more natural to have a post-has-tag condition be represented like `{Post} {Has} {Tag}` instead of `{Post} {Tags} {Contains} >> {ID} {Equals}` (although the latter is more flexible). And a post-type-is condition to be represented like `{Post} {Type} {Is}` rather than `{Post} {Type} {Name} {Equals}`.

And that it the reason why we are here. 

---

> One possible solution for this would be to also show the ID of the entity in the dropdown (like Human ID (ID)).

This would be useful when there are duplicate Human IDs though (edge case alert!).

**Comment by JDGrimes** [comment]
> So, for the user, it might seem more natural to have a post-has-tag condition be represented like {Post} {Has} {Tag} instead of {Post} {Tags} {Contains} >> {ID} {Equals} (although the latter is more flexible). And a post-type-is condition to be represented like {Post} {Type} {Is} rather than {Post} {Type} {Name} {Equals}.

There are several things we can glean from this:
1. We might want to tweak some of our condition names.
2. We might want to rethink how we handle entity arrays.
3. We might want to use direct identity conditions.

However, we also have to balance ease of use with practicability. And we need to consider whether we can sacrifice parts of a feature to make the UI simpler as a whole; we should consider what we'd have to trade for these (supposed) gains. And that is why we _might_ want to do some of those things, while others we _might not_.

### Tweaking Condition Names

This I think would be a good idea. But I'm not sure what we would do with the Contains condition if we keep entity arrays around. Perhaps "Matches" or "Includes" could be good names, but I think we might need to rethink that condition anyway.

### Direct Identity Conditions

This also seems like a good idea, but we have to consider how we'll do it. There are several options we've explored above, and none of them is perfect.

### Rethinking Entity Array Handling

Completely rethinking the way that we handle entity arrays would be a major change. But it is also a big part of why I'm so unsatisfied with this UI right now.

On further consideration, the `{Entity} {Has} {Child}` format laid out above seems a little weird. `{Entity} {Has [Child]}` might seem a bit more natural. I guess we could even do `{Entity} {Has Child} {Child}`, but using the term "child" seems a little odd.

Of course, all of these arrangements basically imply that we'll be dealing with the same dropdown issues as with the direct identity conditions (which these are basically one form of).

**Comment by JDGrimes** [comment]
> ## UI: Direct Identity Conditions

Another way to handle this:

### Single Dropdown for Entities and Children

Instead of displaying the main args in one dropdown, and the children in another one once an arg is selected, we could display all args and children (to arbitrary depth) in a single dropdown.

#### Problems
- The dropdown would take more work to build (i.e., all the work would be done at once).
- The dropdown could become very voluminous.
- The dropdown could also become quite large in terms of length in characters/pixels.

This does avoid all of the problems with the other options. However, I'm not sure whether the benefits outweigh these costs.

There are probably ways that we could avoid the latter two problems, and it is possible that in the process the first could also be mitigated. The work involved in building the dropdown would probably not be a problem unless there were an unrealistically huge number of args. And if it does become an issue, we can burn that bridge when we come to it.

One part of this that we could look at as a positive is that it would require less clicks and selections to add a condition.

I think maybe I'll give this a try and just see how it works.

**Comment by JDGrimes** [comment]
> I think maybe I'll give this a try and just see how it works.

The single dropdown simplifies and unites the UI, and also makes the code much simpler as well. I have not yet implemented the identity conditions, but that is the next step, and I think these are steps in the right direction.

**Comment by JDGrimes** [comment]
We now have support for Identity Conditions in both the UI and the API. There are probably refinements still to be made in both areas, but I'm going to call this fixed for now.
